Fix assertion failure in pytorch profiler

kadeng opened the pull request:
Summary:
Attempt to fix the following exception which occurred when profiling a Pytorch model ( Meta-internal LLM ) that also involved a ThreadPoolExecutor in the background:
```
Exception Found: !stack.empty() INTERNAL ASSERT FAILED at "fbcode/caffe2/torch/csrc/autograd/profiler_python.cpp":987, please report a bug to PyTorch. Python replay stack is empty.
```
The root cause of this issue seems to be that a thread call stack can be empty, which is asserted to not be empty.

I fixed this with some minimal changes to profiler_python.cpp

Approach: 
 * Ensuring that the stack in question is not empty before trying to pop from it.

Test Plan:
* Tested manually on a reproducible scenario where the assertion failure was otherwise triggered ( repro too large to include here ). The assertion failure disappears.
 * CI

Differential Revision: D67691558




cc @robieta @chaekit @guotuofeng @guyang3532 @dzhulgakov @davidberard98 @briancoutinho @sraikund16 @sanrise

kadeng commented:
@sraikund16 Before merging, it would be great to also get feedback from you. I am not 100% sure if acquiring the GIL is strictly neccessary in this case. I am not familiar with the torch profiler codebase, so it's unclear to me whether these call stacks are live python interpreter internal data structures or not, and/or if the code section is already under GIL.

sraikund16 commented:
> @sraikund16 Before merging, it would be great to also get feedback from you. I am not 100% sure if acquiring the GIL is strictly neccessary in this case. I am not familiar with the torch profiler codebase, so it's unclear to me whether these call stacks are live python interpreter internal data structures or not, and/or if the code section is already under GIL.

Added comment internally. TL;DR, GIL should already be acquired. Please update PR to remove that call

kadeng commented:
I removed the line that acquired the GIL....

kadeng committed:
Fix assertion failure in pytorch profiler (#143940)

Summary:

Attempt to fix the following exception which occurred when profiling a Pytorch model ( Meta-internal LLM ) that also involved a ThreadPoolExecutor in the background:
```
Exception Found: !stack.empty() INTERNAL ASSERT FAILED at "fbcode/caffe2/torch/csrc/autograd/profiler_python.cpp":987, please report a bug to PyTorch. Python replay stack is empty.
```
The root cause of this issue seems to be that a thread call stack can be empty, which is asserted to not be empty. 

I fixed this with some minimal changes to profiler_python.cpp

Approach: 
 * Ensuring that the stack in question is not empty before trying to pop from it.

Test Plan:
* Tested manually on a reproducible scenario where the assertion failure was otherwise triggered ( repro too large to include here ). The assertion failure disappears.
 * CI

Reviewed By: sraikund16

Differential Revision: D67691558
